### PR TITLE
[per-OS Packages] Add platforms and excluded_platforms functionality

### PR DESCRIPTION
**Pull request by savil**
## Summary

**Motivation**
Some users have a need to install platform-specific packages. For example, `glibcLocales` cannot be installed on macOS but may be needed on Linux.

**Approach**
We enable two flags in `devbox add`:
```
# install on all platforms except x86_64-darwin
> devbox add glibcLocales --exclude-platform x86_64-darwin

# only install on x86_64-darwin and aarch64-darwin
> devbox add darwin.apple_sdk.frameworks.Security --platform x86_64-darwin 
> devbox add darwin.apple_sdk.frameworks.Security --platform aarch64-darwin 
```

Follow up PRs:
- [ ]  enable multiple `--platform` flag uses
- [ ] detect if `devbox add` fails due to platform-mismatch and suggest `--exclude-platform` flag

## How was it tested?

Added testscript unit test

Adhoc testing:
```
savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> rm -rf .devbox
savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> devbox shell
Ensuring packages are installed.

Installing package: go@1.19.8.

[1/1] go@1.19.8
[1/1] go@1.19.8: Success
Starting a devbox shell...
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
```

Install for a different platform than the one i am currently running on
```
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> devbox add hello --platform x86_64-linux
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> cat devbox.json
{
  "packages": {
    "go": "1.19.8",
    "hello": {
      "version": "latest",
      "platforms": [
        "x86_64-linux"
      ]
    }
  },
  "shell": {
    "init_hook": [
      "export \"GOROOT=$(go env GOROOT)\""
    ],
    "scripts": {
      "run_test": "go run main.go"
    }
  }
}
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> which hello
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2) [1]>
```

Install on all platforms except my current platform
```
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2) [1]> devbox add cowsay --exclude-platform x86_64-darwin
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> which cowsay
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2) [1]> cat devbox.json
{
  "packages": {
    "go": "1.19.8",
    "hello": {
      "version": "latest",
      "platforms": [
        "x86_64-linux"
      ]
    },
    "cowsay": {
      "version": "latest",
      "excluded_platforms": [
        "x86_64-darwin"
      ]
    }
  },
  "shell": {
    "init_hook": [
      "export \"GOROOT=$(go env GOROOT)\""
    ],
    "scripts": {
      "run_test": "go run main.go"
    }
  }
}
```

Ensure basic adding still works
```
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> which vim
/usr/bin/vim
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> devbox add vim

Installing package: vim@latest.

[1/1] vim@latest
[1/1] vim@latest: Success
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> which vim
/Users/savil/code/jetpack/devbox/examples/development/go/hello-world/.devbox/virtenv/.wrappers/bin/vim
(devbox) savil@Savil-Srivastavas-MacBook-Pro ~/c/j/d/e/d/g/hello-world (savil/config-packages-2)> cat devbox.json
{
  "packages": {
    "go": "1.19.8",
    "hello": {
      "version": "latest",
      "platforms": [
        "x86_64-linux"
      ]
    },
    "cowsay": {
      "version": "latest",
      "excluded_platforms": [
        "x86_64-darwin"
      ]
    },
    "vim": "latest"
  },
  "shell": {
    "init_hook": [
      "export \"GOROOT=$(go env GOROOT)\""
    ],
    "scripts": {
      "run_test": "go run main.go"
    }
  }
}
```

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1357** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1357" target="_blank">Graphite</a> 
    * **PR #1358** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1358" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/1358?utm_source=stack-comment).

**Comment by savil** [comment]
On second thought, this needs a fix: we should update the lockfile, even if not installing on the user's platform.

File: internal/devconfig/packages.go [hunk 1]
```diff
@@ -55,6 +59,84 @@ func (pkgs *Packages) Remove(versionedName string) {
...
+				if existing == platform {
+					alreadyPresent = true
+					break
+				}
+			}
+
+			// Add the platform if it's not already present
+			if !alreadyPresent {
+				pkg.Platforms = append(pkg.Platforms, platform)
+			}
```
Comment (savil) [inline internal/devconfig/packages.go:84]: this is verbose I realize. In the next PR, it gets simplified to:
```
if !lo.SomeBy(pkg.Platforms, func(p string) bool { return p == platform }) {
  pkg.Platforms = append(pkg.Platforms, platform)
}
```

File: internal/boxcli/add.go [hunk 2]
```diff
@@ -18,8 +18,10 @@ import (
 const toSearchForPackages = "To search for packages, use the `devbox search` command"
 
 type addCmdFlags struct {
-	config        configFlags
-	allowInsecure bool
+	config          configFlags
+	allowInsecure   bool
+	platform        string
+	excludePlatform string
```
Comment (mikeland73) [inline internal/boxcli/add.go:24]: Should be slice?

File: internal/impl/packages.go [hunk 3]
```diff
@@ -86,6 +87,19 @@ func (d *Devbox) Add(ctx context.Context, pkgsNames ...string) error {
...
+				return err
+			}
+		}
+		if excludePlatform != "" {
+			if err := d.cfg.Packages.ExcludePlatform(pkg, excludePlatform); err != nil {
+				return err
+			}
+		}
+	}
+
```
Comment (mikeland73) [inline internal/impl/packages.go:102]: Can this be part of `d.cfg.Packages.Add(packageNameForConfig)` ?

File: internal/devconfig/packages.go [hunk 4]
```diff
@@ -143,13 +226,45 @@ func NewPackage(name string, values map[string]any) Package {
...
+		Platforms:         platforms,
+		ExcludedPlatforms: excludedPlatforms,
 	}
 }
 
+// enabledOnPlatform returns whether the package is enabled on the given platform.
+// If the package has a list of platforms, it is enabled only on those platforms.
+// If the package has a list of excluded platforms, it is enabled on all platforms
+// except those.
+func (p *Package) enabledOnPlatform(platform string) bool {
```
Comment (mikeland73) [inline internal/devconfig/packages.go:251]: can this just be part of `IsEnabledOnPlatform` ?

File: internal/nix/nix.go [hunk 5]
```diff
@@ -169,6 +170,29 @@ func Version() (string, error) {
...
+	// not technically supported, but should work?
+	// ref. https://nixos.wiki/wiki/Nix_on_ARM
+	// ref. https://github.com/jetpack-io/devbox/pull/1300
+	"armv7l-linux",
+}
+
+// EnsureValidPlatform returns an error if the platform is not supported by nix.
+// https://nixos.org/manual/nix/stable/installation/supported-platforms.html
+func EnsureValidPlatform(platform string) error {
+	for _, p := range nixPlatforms {
```
Comment (mikeland73) [inline internal/nix/nix.go:188]: `lo.Contains` ? or just use a map.

File: internal/impl/packages.go [hunk 6]
```diff
@@ -221,6 +235,17 @@ func (d *Devbox) ensurePackagesAreInstalled(ctx context.Context, mode installMod
...
+	// Update lockfile with new packages that are not to be installed
+	cfgPkgs, err := d.ConfigPackages()
+	if err != nil {
+		return err
+	}
+	for _, pkg := range cfgPkgs {
+		if err := pkg.EnsureUninstallableIsInLockfile(); err != nil {
+			return err
+		}
+	}
```
Comment (mikeland73) [inline internal/impl/packages.go:247]: EnsureUninstallableIsInLockfile is a bit awkward, you could simply loop on `d.ConfigPackages()` and resolve all.

If they are already resolved, it's a no-op.

**Review by mikeland73 — APPROVED** [review]

File: internal/impl/packages.go (hunk 6 above)
Comment (savil) [reply internal/impl/packages.go:254]: it has some more checks inside the function....

File: internal/devconfig/packages.go (hunk 4 above)
Comment (savil) [reply internal/devconfig/packages.go:251]: lol, yes

File: internal/boxcli/add.go (hunk 2 above)
Comment (savil) [reply internal/boxcli/add.go:24]: yup, done in next PR

File: internal/impl/packages.go (hunk 3 above)
Comment (savil) [reply internal/impl/packages.go:102]: This is a good idea. I'll do it in a stack PR (#1363) to minimize conflicts.

File: internal/nix/nix.go (hunk 5 above)
Comment (savil) [reply internal/nix/nix.go:196]: I find current version simpler to write and read

File: internal/impl/packages.go (hunk 3 above)
Comment (savil) [reply internal/impl/packages.go:102]: actually, moving this won't work.

We need this here to enable the following scenario:
```
devbox add hello --platform x86_64-linux
devbox add hello --platform x86-64-darwin
```

The second time it runs, it willl not be within the loop above where .Add() is called.

**Comment by savil** [comment]
test failure is spurious. Tests pass in subsequent PRs.

**Pull request merged**